# Scraping Stackoverflow Questions (test 2)

- Get question, views, vote-count, and question link

In [1]:
import requests
from bs4 import BeautifulSoup
import json

# Question links

In [3]:
questions_data = {
    "questions": []
}


for i in range(2):
    
    link = 'https://stackoverflow.com/questions?tab=frequent&page=' + str(i)
    
    print(link)
    res = requests.get(link)
    soup = BeautifulSoup(res.text, "html.parser")
    questions = soup.select(".question-summary")
    
    for que in questions:
        q = que.select_one('.question-hyperlink').getText()
        vote_count = que.select_one('.vote-count-post').getText()
        views = que.select_one('.views').attrs['title']
        questions_data['questions'].append({
            "question": q,
            "views": views,
            "vote_count": vote_count
        })

https://stackoverflow.com/questions?tab=frequent&page=0
https://stackoverflow.com/questions?tab=frequent&page=1


In [4]:
questions_data['questions']

[{'question': 'What is a NullPointerException, and how do I fix it?',
  'views': '2,872,497 views',
  'vote_count': '209'},
 {'question': 'How to make a great R reproducible example',
  'views': '318,101 views',
  'vote_count': '2472'},
 {'question': 'How do I return the response from an asynchronous call?',
  'views': '1,522,378 views',
  'vote_count': '5653'},
 {'question': 'How can I prevent SQL injection in PHP?',
  'views': '1,786,579 views',
  'vote_count': '2773'},
 {'question': 'RegEx match open tags except XHTML self-contained tags',
  'views': '2,942,309 views',
  'vote_count': '1554'},
 {'question': 'What is a NullReferenceException, and how do I fix it?',
  'views': '1,426,307 views',
  'vote_count': '1874'},
 {'question': 'How do I compare strings in Java?',
  'views': '3,938,942 views',
  'vote_count': '724'},
 {'question': '“Notice: Undefined variable”, “Notice: Undefined index”, and “Notice: Undefined offset” using PHP',
  'views': '1,794,455 views',
  'vote_count': '12

In [5]:
questions_data = {
    "questions": []
}

link = 'https://stackoverflow.com/questions?tab=frequent&page=' + str(1)
    
res = requests.get(link)
soup = BeautifulSoup(res.text, "html.parser")
questions = soup.select(".question-summary")

for que in questions:
    q = que.select_one('.question-hyperlink').getText()
    vote_count = que.select_one('.vote-count-post').getText()
    views = que.select_one('.views').attrs['title']
    questions_data['questions'].append({
        "question": q,
        "views": views,
        "vote_count": vote_count
    })

https://stackoverflow.com/questions?tab=frequent&page=1


In [6]:
questions

[<div class="question-summary" id="question-summary-218384">
 <div class="statscontainer">
 <div class="stats">
 <div class="vote">
 <div class="votes">
 <span class="vote-count-post"><strong>209</strong></span>
 <div class="viewcount">votes</div>
 </div>
 </div>
 <div class="status answered-accepted">
 <strong>12</strong>answers
             </div>
 </div>
 <div class="views supernova" title="2,872,498 views">
     2.9m views
 </div>
 </div>
 <div class="summary">
 <h3><a class="question-hyperlink" href="/questions/218384/what-is-a-nullpointerexception-and-how-do-i-fix-it">What is a NullPointerException, and how do I fix it?</a></h3>
 <div class="excerpt">
             What are Null Pointer Exceptions (java.lang.NullPointerException) and what causes them?
 
 What methods/tools can be used to determine the cause so that you stop the exception from causing the program ...
         </div>
 <div class="tags t-java t-nullpointerexception">
 <a class="post-tag" href="/questions/tagged/java"

In [12]:
for que in questions:
    q = que.select_one('.question-hyperlink').attrs['href']
    
    link = 'https://stackoverflow.com'+ q
    print(link)

https://stackoverflow.com/questions/218384/what-is-a-nullpointerexception-and-how-do-i-fix-it
https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example
https://stackoverflow.com/questions/14220321/how-do-i-return-the-response-from-an-asynchronous-call
https://stackoverflow.com/questions/60174/how-can-i-prevent-sql-injection-in-php
https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags
https://stackoverflow.com/questions/4660142/what-is-a-nullreferenceexception-and-how-do-i-fix-it
https://stackoverflow.com/questions/513832/how-do-i-compare-strings-in-java
https://stackoverflow.com/questions/4261133/notice-undefined-variable-notice-undefined-index-and-notice-undefined
https://stackoverflow.com/questions/588004/is-floating-point-math-broken
https://stackoverflow.com/questions/388242/the-definitive-c-book-guide-and-list
https://stackoverflow.com/questions/12859942/why-shouldnt-i-use-mysql-functions-in-php
https://stac

In [14]:
questions_data = {
    "questions": []
}

link = 'https://stackoverflow.com/questions?tab=frequent&page=' + str(1)
    
res = requests.get(link)
soup = BeautifulSoup(res.text, "html.parser")
questions = soup.select(".question-summary")

for que in questions:
    q = que.select_one('.question-hyperlink').getText()
    question_link = 'https://stackoverflow.com'+ que.select_one('.question-hyperlink').attrs['href']
    vote_count = que.select_one('.vote-count-post').getText()
    views = que.select_one('.views').attrs['title']
    questions_data['questions'].append({
        "question": q,
        "views": views,
        "vote_count": vote_count,
        "question_link": question_link
    })

In [19]:
questions_data['questions'][0]

{'question': 'What is a NullPointerException, and how do I fix it?',
 'views': '2,872,498 views',
 'vote_count': '209',
 'question_link': 'https://stackoverflow.com/questions/218384/what-is-a-nullpointerexception-and-how-do-i-fix-it'}

In [46]:
#link = 'https://stackoverflow.com/questions?tab=frequent&page=' + str(i)
    
#res = requests.get(link)
#soup = BeautifulSoup(res.text, "html.parser")

In [47]:
#soup

# Answers in the link

In [20]:
question_link = 'https://stackoverflow.com/questions/218384/what-is-a-nullpointerexception-and-how-do-i-fix-it'

In [30]:
res = requests.get(question_link)

In [44]:
soup = BeautifulSoup(res.text, "html.parser")
#answers = soup.select(".question")
#answers = soup.find("div", {"id": "answers"})

answers = soup.find("div", {"class": "answer accepted-answer"})

In [45]:
answers

<div class="answer accepted-answer" data-answerid="218510" id="answer-218510" itemprop="acceptedAnswer" itemscope="" itemtype="http://schema.org/Answer">
<div class="post-layout">
<div class="votecell post-layout--left">
<div class="js-voting-container grid fd-column ai-stretch gs4 fc-black-200" data-post-id="218510">
<button aria-label="Up vote" aria-pressed="false" class="js-vote-up-btn grid--cell s-btn s-btn__unset c-pointer" data-controller="s-tooltip" data-s-tooltip-placement="right" data-selected-classes="fc-theme-primary" title="This answer is useful"><svg aria-hidden="true" class="m0 svg-icon iconArrowUpLg" height="36" viewbox="0 0 36 36" width="36"><path d="M2 26h32L18 10 2 26z"></path></svg></button>
<div class="js-vote-count grid--cell fc-black-500 fs-title grid fd-column ai-center" data-value="3807" itemprop="upvoteCount">3807</div>
<button aria-label="Down vote" aria-pressed="false" class="js-vote-down-btn grid--cell s-btn s-btn__unset c-pointer" data-controller="s-tooltip

In [50]:
answers.select_one('.s-prose js-post-body').getText()

AttributeError: 'NoneType' object has no attribute 'getText'

### TODO:

How to scrape the accepta answer, from 

div class="answer accepted-answer" data-answerid="218510" id="answer-218510" itemprop="acceptedAnswer" itemscope="" itemtype="http://schema.org/Answer" 

to div class="s-prose js-post-body" itemprop="text"?


In [55]:
answers.select_one('.js-voting-container grid fd-column ai-stretch gs4 fc-black-200')

In [58]:
soup.find("div", {"class": "answer accepted-answer"}).find("div", {"class": "s-prose js-post-body"}).getText()

'\nWhen you declare a reference variable (i.e. an object) you are really creating a pointer to an object. Consider the following code where you declare a variable of primitive type int:\nint x;\nx = 10;\n\nIn this example, the variable x is an int and Java will initialize it to 0 for you. When you assign it the value of 10 on the second line, your value of 10 is written into the memory location referred to by x.\nBut, when you try to declare a reference type, something different happens. Take the following code:\nInteger num;\nnum = new Integer(10);\n\nThe first line declares a variable named num, but it does not actually contain a primitive value yet. Instead, it contains a pointer (because the type is Integer which is a reference type). Since you have not yet said what to point to, Java sets it to null, which means "I am pointing to nothing".\nIn the second line, the new keyword is used to instantiate (or create) an object of type Integer, and the pointer variable num is assigned to 